docs/design-proposals/global-scheduler/GlobalSchedulerArchitectureDesignSpecificationFirstDraft.md

```diff
@@ -0,0 +1,416 @@
+
+- shared state lock-free optimistic concurrent scheduling operation
+- multi-dimensional optimization-based scheduling algorithm
+- dynamic geolocation and resource profile-based partition scheme
+- conflict aware VM/Container request distribution and scheduling mechanism
+- smart conflict avoidance approach
+
+## 2.3 High-Level Architecture Design
+
+We divide the scheduling into two levels. The first level Global Scheduler selects the best cluster for the VM/Container request. We then leverage the existing cluster level scheduler to select the best node that will host and run the VM/Container. Our Two-Level Scheduler Architecture is different from what is described in section 1.3.6 since our first level Global Scheduler is not a single monolithic scheduler. It employs a distributed model that consists of multiple Global Schedulers running in a lock-free concurrent mode, thus our new design avoids the issues of scalability bottleneck and low allocation rate. The architecture allows flexible scale-out to the needed global capacity.
+
+In our design, all the Global Schedulers operate in a shared-state manner. That is, they shared one global resource DB which holds the cluster-level resource information of all the clusters. The DB size will be manageable since the DB only holds cluster-level resource information, such as available cluster CPU, available cluster memory, supported VM flavor list, and network topology of clusters, network proximity of clusters, etc.. Unlike the Shared-State Multi-Scheduler Architecture outlined in section 1.3.3, the global resource DB is not replicated by each Global Scheduler. Instead, they share and read/update information from one global resource DB.  In our new design, the Global Schedulers run in an optimistic concurrent way instead of using pessimistic concurrency control, which removes the locking latency.
+
+Unlike existing multi-scheduler architecture design as described in section 1.3.2 and 1.3.3, in which the incoming scheduling requests are either distributed to all schedulers or a scheduler in a round-robin way, in our design, the incoming scheduling requests are distributed to a Global Scheduler based on the request's profile. Section 2.5.4 illustrates how this distribution mechanism combined with the Global Scheduler partitioning scheme (Section 2.5.3) will mitigate the scheduling conflict when all the Global Schedulers run concurrently.
+
+In summary, this design eliminates the issues associated with existing Two-Level Scheduler Architecture, Siloed Multi-Scheduler Architecture, and Shared-State Multi-Scheduler Architecture. It renders good scalability, low latency, and global optimal scheduling while satisfying the user-supplied placement constraints.
+
+On the southbound, it integrates with different types of clusters via a flexible plugin design.
+
+Figure.1 is an illustration of the high-level architecture.
+![image.png](/images/2.3.png)
```

Comment:
1. Does the distributor in the figure need to watch vm/container resources through the API server? Is the arrow from the API server to the distributor in reverse?
2. How does the Distributor communicate with the global scheduler?
Distributor and Global Scheduler are different processes, right?
3. Indicates whether the request dispatcher supports batch operations. Currently, batch operations may also be a major requirement scenario.

Reply:
1. yes
2. Good question. This is open for discussion. There are two ways to implement this. 
a) Each scheduler is a separate process and the distributor is another process which interacts with the API Server to get the VM/Container events. The distributor can list&watch VM/Container events with API server through existing K8S "informer" mechanism. The interface between the distributor and each scheduler can be based on the same "informer" mechanism, i.e. each scheduler will register with the distributor a "watch for its associated resource profile tags" and get filtered notification of those VM/Containers that match the registered tags. The interface between the distributor and each scheduler can alternatively be gRPC.  
b) Each scheduler is a thread and has its own scheduling event queue. All schedulers share the same process but have separate scheduling event queues. The distributor runs inside the scheduling process and list&watch VM/Container events with API server through existing K8S "informer" mechanism. Then the distributor puts filtered VM/container requests into each scheduler's queue based on match between the VM/Container's profile and the scheduler queue's profile tags.